File: XTEA/MicroPython/XTEA no OS.py
import machine
import utime
import uos
from xtea_cipher import *

freq = 200000000
printfreq = int(str(freq)[:3])
datasize = 16

plain = ""
plain16 = "sixteenbytetext"
plain32 = "sixteenbytetextsixteenbytestext"
plain64 = "sixteenbytetextsixteenbytestextsixteenbytestextsixteenbytestext"
plain128 = "sixteenbytetextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestext"
plain256 = "sixteenbytetextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestext"
plain512 = "sixteenbytetextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestext"

key = bytearray([
    0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08, 
    0x09, 0x0A, 0x0B, 0x0C, 0x0D, 0x0E, 0x0F, 0x10
])

iv = bytearray([
    0x00, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07
])
encryptedString = ""
encryptStartTime = 0
encryptEndTime = 0
encryptElapsedMicros = 0
decryptStartTime = 0
decryptEndTime = 0
decryptElapsedMicros = 0

machine.freq(freq)
print("No OS || ", printfreq, "mhz || ", "XTEA", " || datasize ", datasize)

def decryptData(encryptedData):
    decipher = aes(key, MODE_CBC, iv)
    decryptedData = decipher.decrypt(encryptedData)
    return decryptedData.strip().decode('utf-8')
   
if datasize == 16:
    plain = plain16
elif datasize == 32:
    plain = plain32
elif datasize == 64:
    plain = plain64
elif datasize == 128:
    plain = plain128
elif datasize == 256:
    plain = plain256
elif datasize == 512:
    plain = plain512

while True:
	data_bytes = bytes(data, 'utf-8')
    cipher = xtea(key, MODE_ECB, iv)
	encryptStartTime = utime.ticks_us()
    encryptedData = cipher.encrypt(data_bytes) 
    encryptEndTime = utime.ticks_us()
    encryptElapsedMicros = encryptEndTime - encryptStartTime
	decipher = xtea(key, MODE_ECB, iv)
	decryptStartTime = utime.ticks_us()
    decryptedData = decipher.decrypt(encryptedData)
    decryptEndTime = utime.ticks_us()
    decryptElapsedMicros = decryptEndTime - decryptStartTime
    
#     print(decryptedData)
    print('E', encryptElapsedMicros, 'D', decryptElapsedMicros)
    utime.sleep_ms(3000)

File: XTEA/C++/XTEA no OS.cpp
#include <Crypto.h>
#include <Arduino.h>
#include "hardware/clocks.h"
#include "pico/bootrom.h"
#include "pico.h"
#include <string.h>
#define XTEA_ROUNDS 48
#define XTEA_MAC_ROUNDS 24
#include "XTEA-Cipher.h"

#define XTEA_BLOCK_SIZE 8
#define XTEA_IV_SIZE 8
#define XTEA_KEY_SIZE 16

unsigned long encryptStartTime;
unsigned long encryptEndTime;
unsigned long encryptElapsedMicros;
unsigned long decryptStartTime;
unsigned long decryptEndTime;
unsigned long decryptElapsedMicros;

uint8_t iv[XTEA_IV_SIZE] = {0x00, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07};

uint8_t key[XTEA_KEY_SIZE] = {
    0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08,
    0x09, 0x0A, 0x0B, 0x0C, 0x0D, 0x0E, 0x0F, 0x10};

int data_size = 16;
char packet[512];
char packet16[] = "sixteenbytetext";
char packet32[] = "sixteenbytetextsixteenbytestext";
char packet64[] = "sixteenbytetextsixteenbytestextsixteenbytestextsixteenbytestext";
char packet128[] = "sixteenbytetextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestext";
char packet256[] = "sixteenbytetextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestext";
char packet512[] = "sixteenbytetextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestextsixteenbytestext";
int packetSize = 0;

void time_output() {
    Serial.print("E ");
    Serial.print(encryptElapsedMicros);
    Serial.print("  D ");
    Serial.println(decryptElapsedMicros);
}

void decrypted_output(uint8_t* buffer, size_t size) {
    Serial.print("Decrypted message: ");
    for (size_t i = 0; i < size; ++i) {
        Serial.print((char)buffer[i]);
    }
    Serial.println();
}

void setup() {
    delay(10000);

    if (data_size == 16) {
        strcpy(packet, packet16);
    } else if (data_size == 32) {
        strcpy(packet, packet32);
    } else if (data_size == 64) {
        strcpy(packet, packet64);
    } else if (data_size == 128) {
        strcpy(packet, packet128);
    } else if (data_size == 256) {
        strcpy(packet, packet256);
    } else if (data_size == 512) {
        strcpy(packet, packet512);
    }

    Serial.begin(9600);
    Serial.println(" ");
    Serial.print("No OS || ");
    Serial.print(clock_get_hz(clk_sys) / 1000000);
    Serial.print("mhz || ");
    Serial.print("XTEA");
    packetSize = strlen(packet);
    Serial.print(" || datasize ");
    Serial.println(packetSize + 1);
}

void loop() {
    uint8_t resultBuffer[data_size];
    memcpy(resultBuffer, packet, data_size);

    encryptStartTime = micros();
    xtea.ecbEncrypt(key, resultBuffer, data_size);
    encryptEndTime = micros();
    encryptElapsedMicros = encryptEndTime - encryptStartTime;

    decryptStartTime = micros();
    xtea.ecbDecrypt(key, resultBuffer, data_size);
    decryptEndTime = micros();
    decryptElapsedMicros = decryptEndTime - decryptStartTime;

    time_output();
    decrypted_output(resultBuffer, data_size);
    delay(3000);
}

